proposals/xxxx-mesh-nodes.md

```diff
@@ -0,0 +1,329 @@
+- The Output signature describes `vertices` output elements
+- The PatchConstantOrPrimitive signature describes `primitives` output elements.
+
+In addition to the intrinsics supported by a broadcast launch node shader with
+no output nodes, the following mesh shader intrinsics are also supported by
+node shaders with `mesh` launch type:
+- `dx.op.setMeshOutputCounts`
+- `dx.op.EmitIndices`
+- `dx.op.StoreVertexOutput`
+- `dx.op.StorePrimitiveOutput`
+
+Notably, the `dx.op.GetMeshPayload` intrinsic from mesh shader is not supported
+by mesh nodes. Instead, payload data is available through the node input record.
+
+### Runtime Additions
+
+#### Runtime information
+
+For mesh launch mode, additional node details will need to be captured to RDAT.
+For context, see [RDAT_LibraryTypes.inl](https://github.com/microsoft/DirectXShaderCompiler/blob/main/include/dxc/DxilContainer/RDAT_LibraryTypes.inl).
+The approach proposed for this prototype is to add a `RDAT::NodeFuncAttribKind`
+of `MeshShaderInfo`, and add an entry to the union:
+`RDAT_RECORD_REF(MSInfo, MeshShaderInfo)`.
+This will use the existing `MSInfo` structure to encode the required details
+for the mesh shader.
+There are a few redundant fields of this structure, given what's already
+defined for the node shader, but for the experimental feature, the trade-off is
+probably worth the simplicity of not adding another record table at this point.
+
+For now, here is a breakdown of the fields in the `MSInfo` structure, and how
+they will be used in a mesh node:
+- `SigOutputElements` - used for the elements of the `vertices` output array.
+- `SigPrimOutputElements` - used for the elements of the `primitives` output array.
+- `ViewIDOutputMask` - unset for now.
+- `ViewIDPrimOutputMask` - unset for now.
+- `NumThreads` - redundant with `NodeShaderFuncAttrib` NumThreads.  Could be used instead of the additional attribute, or we could copy the ref so it's identical in both places.
+- `GroupSharedBytesUsed` - redundant with same field in `NodeShaderInfo`.  Set to same value.
+- `GroupSharedBytesDependentOnViewID` - set to zero for now.
+- `PayloadSizeInBytes` - redundant with input record size info.  We shouldn't use this one.  Set to zero?
+- `MaxOutputVertices` - based on `vertices` array size
+- `MaxOutputPrimitives` - based on `primitives` array size
+- `MeshOutputTopology` - `uint8_t` encoded `DXIL::MeshOutputTopology` based on `[OutputTopology(`*topology*`)]`
+
+#### Device Capability
+
+See the work graphs spec
+[here](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#graphics-nodes)
+for detail.
+
+Devices that support experimental `D3D_SHADER_MODEL_6_9` and experimental [`D3D12_WORK_GRAPHS_TIER_1_1`](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_work_graphs_tier) are required to support these features as part of graphics nodes in work graphs.
+
+## Open Questions
+
+- Should vanilla mesh shaders be supported, or only mesh nodes?
```

Comment:
I don't know what this implies. can you elaborate?

Reply:
This is asking if the path the initial prototype allowed should go forward -> allowing existing mesh shaders to be used, with payload acting as node input record.   I think the answer here has to be no, since existing mesh shaders don't include some of the metadata that the runtime needs in order to enable all ways of connecting into a graph.  For instance there isn't metadata indicating where SV_DispatchGrid is located in the input record, which the runtime needs to tell driver how to handle entrypoint input to a node, so existing mesh shaders can't be used at the entry of a graph.  Might as well keep it simple and only support mesh nodes.  Similarly we kept it simple and didn't expose vanilla compute shaders.